src/lib.rs

```diff
@@ -196,14 +196,66 @@ macro_rules! custom_error {
                 match self {$(
                     $errtype::$field $( { $( $attr_name ),* } )* => {
                         $(write!(formatter, "{}", ($($msg_fun)*) )?;)*
-                        $crate::display_message!(formatter, $($($attr_name),*),* | $($msg)*);
+                        $crate::display_message_for_enum!(formatter, $($($attr_name),*),* | $($msg)*);
                         Ok(())
                     }
                 ),*}
             }
         }
         }}
     };
+    (
+        $( ($prefix:tt) )* // `pub` marker
+        $errtype:ident // Name of the error type to generate
+        $( < $(
+            $type_param:tt // Optional type parameters for generic error types
+            ),*
+        > )*
+        { $(
+            $field_name:ident // Name of an attribute of the error variant
+            :
+            $($field_type:ident)::* // type of the attribute
+            $(< $($field_type_param:tt),* >)* // Generic (lifetime & type) parameters for the attribute's type
+        ),* }
+        =
+        $( @{ $($msg_fun:tt)* } )*
+        $($msg:expr)* // The human-readable error message
+        $(,)* // Trailing comma
+    ) => {
+        #[derive(Debug)]
+        $($prefix)* struct $errtype $( < $($type_param),* > )* {
+            $( $field_name : $($field_type)::* $(< $($field_type_param),* >)* ),*
+        }
+
+        $crate::add_type_bounds! {
+        ( $($($type_param),*)* )
+        (std::fmt::Debug + std::fmt::Display)
+        { impl <} {> std::error::Error
+            for $errtype $( < $($type_param),* > )*
+        {
+            fn source(&self) -> Option<&(dyn std::error::Error + 'static)>
+            {
+                None
+            }
+        }
+        }}
+
+        $crate::add_type_bounds! {
+        ( $($($type_param),*)* )
+        (std::string::ToString)
+        { impl <} {> std::fmt::Display
+            for $errtype $( < $($type_param),* > )*
+        {
+            fn fmt(&self, formatter: &mut std::fmt::Formatter)
+                -> std::fmt::Result
+            {
+                $(write!(formatter, "{}", ($($msg_fun)*) )?;)*
+                $crate::display_message_for_struct!(self, formatter, $($field_name),* | $($msg)*);
```

Comment:
I think you could reuse the existing `display_message!` macro by matching against self. This would spare us one macro.

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d1c255bdcf0b629b1f468f3b3a67eda4